service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CompleteTableCopyJobsStep.java

```diff
@@ -0,0 +1,100 @@
+import com.google.api.services.bigquery.model.Job;
+import com.google.api.services.bigquery.model.JobReference;
+import com.google.api.services.bigquery.model.TableReference;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CompleteTableCopyJobsStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(CompleteTableCopyJobsStep.class);
+  private final CrlService crlService;
+
+  public CompleteTableCopyJobsStep(CrlService crlService) {
+    this.crlService = crlService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    final Map<String, String> tableToJobId =
+        workingMap.get(ControlledResourceKeys.TABLE_TO_JOB_ID_MAP, new TypeReference<>() {});
+    final AuthenticatedUserRequest userRequest =
+        flightContext
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final Bigquery bigQueryClient = crlService.createNakedBigQueryClient(userRequest);
+    try {
+      for (Map.Entry<String, String> entry : tableToJobId.entrySet()) {
+        final TableReference tableReference = tableFqIdToReference(entry.getKey());
+        final JobReference jobReference = jobFqIdToReference(entry.getValue());
+
+        // wait for job to complete
+        while (true) {
+          final Job job =
+              bigQueryClient
+                  .jobs()
+                  .get(jobReference.getProjectId(), jobReference.getJobId())
+                  .setLocation(
+                      jobReference.getLocation()) // returns NOT_FOUND unless location is specified
+                  .execute();
+          final String jobState = job.getStatus().getState();
+          logger.debug("Table {} is {}", tableReference.getTableId(), jobState);
+          if ("DONE".equals(jobState)) {
+            break;
+          }
+          TimeUnit.SECONDS.sleep(5);
+        }
+      }
+    } catch (IOException | InterruptedException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  // Nothing to undo here because the whole dataset will be deleted in the undo path for
+  // an earlier step.
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
+  }
+
+  // Fully-qualified table IDs are of the form project-id:dataset_id.table_id
+  private TableReference tableFqIdToReference(String tableFqId) {
+    final TableReference result = new TableReference();
+    final String[] outerGroups = tableFqId.split(":");
+    result.setProjectId(outerGroups[0]);
+
+    final String[] innerGroups = outerGroups[1].split("\\.");
```

Comment:
Might be worth confirming in a comment here that GCP disallows ":" and "." characters in names for all of projects, datasets, and tables.
(From what I can tell they do, though a lot more than I would expect is allowed in table names: https://cloud.google.com/bigquery/docs/tables#table_naming)